op/main.py

```diff
@@ -163,8 +198,9 @@ def main(charm_class):
         # When a charm is force-upgraded and a unit is in an error state Juju does not run upgrade-charm and
         # instead runs the failed hook followed by config-changed. Given the nature of force-upgrading
         # the hook setup code is not triggered on config-changed.
-        if (juju_event_name in ('install', 'upgrade-charm') or juju_event_name.endswith('-storage-attached')):
-            _setup_hooks(charm_dir, charm)
+        # 'start' event is included as Juju does not fire the install event for K8s charms (see LP: #1854635).
+        if juju_event_name in ('install', 'start', 'upgrade-charm') or juju_event_name.endswith('-storage-attached'):
```

Comment:
Instead of trying to catch all of the cases where a hook other than `install` is the first thing to run, and potentially getting it wrong, I wonder if we couldn't use our persistence layer to have a flag as to whether we've done the initial setup and then do it when the flag is not set (which is guaranteed to be the first run, regardless of what the actual hook is) and on `upgrade-charm` (which is always going to be a special case, and we might want to handle slightly differently anyway at some point).

Reply:
I think the approach with the persistence layer goes against the general desire to have idempotent hook executions where we evaluate the environment in order to make decisions about what to do next without relying too much on persistent state. Performance-wise this doesn't look like a problem. So I would rather keep this logic in the code itself.

When it comes to `upgrade-charm`, we always have to have it here as there may be new endpoints, storage or core events added for which we will need to create new symlinks (until Juju implements a single entrypoint file feature).